### PR TITLE
20 Implement `Identity.Register` SDK Function

### PR DESCRIPTION
**Pull request by LuisOsta**
Do y'all think the name communicates what it should? What this function does is associate the user's DID with their account in the organization's auth management service. I called it `Register` because its registering the identity in auth management and I think that would be the most clear and explicit to end-user developers consuming the SDK.

### CODE REVIEW

**Comment by LuisOsta** [comment]
In order to actually use this function we'll need to implement the `Token.Create` function described here - https://github.com/knox-networks/knox-go/issues/8

File: service/auth_client/auth_client.go [hunk 1]
```diff
@@ -60,6 +66,25 @@ func (r *authClient) AuthnWithDid(did string, nonce string, enc []byte) error {
 	return nil
 }
 
+func (r *authClient) CreateDidRegistrationChallenge(auth_token string) (*DidRegistrationChallenge, error) {
```
Comment (echenknox) [inline service/auth_client/auth_client.go:69]: Not sure how this is going to work? When you start, it is a stream. It is waiting for 2 responses. One is a challenge, then wait for another one. If you just wrap it up in this call, then the stream will just timeout.

File: identity/identity.go [hunk 2]
```diff
@@ -30,8 +28,35 @@ func NewIdentityClient(address string, s signer.DynamicSigner) (IdentityClient,
...
+
+	err = c.auth.AuthnWithDidRegister(did, nonce, signed)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *identityClient) parseChallenge(challenge params.RegisterIdentityChallenge, token string) (string, error) {
+	if (challenge != params.RegisterIdentityChallenge{}) {
```
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity.go:51]: @echenknox So how would you recommend I determine if a parameter was set or not?

File: identity/identity_test.go [hunk 3]
```diff
@@ -72,6 +76,165 @@ func TestGenerateIdentity(t *testing.T) {
 	}
 }
 
+type registerIdentityields struct {
```
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity_test.go:79]: Yes

File: identity/identity.go [hunk 4]
```diff
@@ -30,8 +28,35 @@ func NewIdentityClient(address string, s signer.DynamicSigner) (IdentityClient,
...
+	} else {
+		challenge, err := c.auth.CreateDidRegistrationChallenge(token)
+		if err != nil {
+			return "", err
+		}
+		return challenge.Nonce, nil
+	}
 }
 
 func (c *identityClient) Generate(params params.GenerateIdentityParams) (*model.DidDocument, *crypto.KeyPairs, error) {
```
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity.go:62]: Yeah I'm aware but I think its better to return an empty struct than nil, I don't think the memory allocation makes much difference

File: identity/identity.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity.go:51]: The brackets are needed in order to do the comparison btw

File: identity/identity.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity.go:62]: For internal things (like the parameters, challenges, etc.) I'm having with nil values. But for external things like what we return, I think we should remove foot guns from the user by returning an empty object instead

File: identity/identity.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity.go]: I've changed this to be pointer based so we can do a nil check

File: identity/identity.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity.go:62]: Also can you be a bit more specific on what you're talking about

File: service/auth_client/auth_client.go (hunk 1 above)
Comment (LuisOsta) [reply service/auth_client/auth_client.go]: Done

File: identity/identity.go [hunk 5]
```diff
@@ -30,11 +28,50 @@ func NewIdentityClient(address string, s signer.DynamicSigner) (IdentityClient,
...
+		}
+		nonce := challenge.Nonce
+		signed, err := c.s.Sign(signer.Authentication, []byte(did+"."+nonce))
+		if err != nil {
+			return err
+		}
+
+		err = c.auth.AuthnWithDidRegister(did, nonce, signed)
+		if err != nil {
+			return err
```
Comment (sKudryashov) [inline identity/identity.go:57]: Would be great to wrap these errors with some messages to be able to track down what went wrong. 
Moreover - are they retryable?

File: identity/identity.go [hunk 6]
```diff
@@ -30,11 +28,50 @@ func NewIdentityClient(address string, s signer.DynamicSigner) (IdentityClient,
...
+	did := c.s.GetDid()
+	if p.Challenge != nil {
+		nonce := p.Challenge.Nonce
+		signed, err := c.s.Sign(signer.Authentication, []byte(did+"."+nonce))
+		if err != nil {
+			return err
+		}
+
+		err = c.auth.AuthnWithDidRegister(did, nonce, signed)
+		if err != nil {
```
Comment (sKudryashov) [inline identity/identity.go:41]: `if err := c.auth.AuthnWithDidRegister(did, nonce, signed); err != nil {`

**Review by sKudryashov — CHANGES_REQUESTED** [review]

File: identity/identity.go (hunk 5 above)
Comment (LuisOsta) [reply identity/identity.go:57]: @sKudryashov I think the errors should be defined by the API and then the SDK should only do minimal changes (if any) to the error, what would you suggest?

File: identity/identity.go (hunk 6 above)
Comment (LuisOsta) [reply identity/identity.go:41]: Sounds good

Comment (LuisOsta) [reply identity/identity.go]: I've refactored the lines to follow this form

File: identity/identity.go (hunk 5 above)
Comment (LuisOsta) [reply identity/identity.go:56]: I think what I'd be okay with doing is to, in a future issue, create a new error object that has both a code and a message (similarly to the gRPC error code).

What do you think?

Comment (sKudryashov) [reply identity/identity.go:56]: I think it can be solved by error types or error interfaces, not error codes.

Comment (adewaleafolabi) [reply identity/identity.go:56]: @LuisOsta I agree the SDK is a wrapper around the core. The improvements to error handling should happen on the API layer.

**Comment by adewaleafolabi** [comment]
IMO Identity.Register seems clear enough. Asides the changes requested, this looks good to me

**Review by sKudryashov — APPROVED** [review]

File: identity/identity.go (hunk 5 above)
Comment (LuisOsta) [reply identity/identity.go:56]: Resolving this now, though lets definitely investigate in the future how we can improve the error handling experience for consumers of the SDK.

Comment (LuisOsta) [reply identity/identity.go:56]: That'll be handled in https://github.com/knox-networks/knox-go/issues/26

**Review by adewaleafolabi — APPROVED** [review]

File: identity/identity.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (LuisOsta) [reply identity/identity.go]: I'm resolving this here but if you have any other issues we can discuss them offline

**Pull request merged**